### DISCUSSION TITLE
How can rest-client reactive follow redirects?

### DISCUSSION THREAD

**Discussion by alex-martel** [quarkusio/quarkus#24989] [Q&A]
I am probably doing it wrong, but I am unable to follow redirects using the rest-client reactive. I scavenged the issue tracker, but just saw that it should be working since quarkus 2.0.

I have set the following in my config; `quarkus.rest-client.rest-api.follow-redirects=true`
I have a mock server returning a 301 with a `Location:` header.

Each time I call it I get an exception;
` org.jboss.resteasy.reactive.ClientWebApplicationException: Received: 'Server response is: 301' when invoking: Rest Client method: 'org.acme.rest-api#test'`

I tried having my own ResponseExceptionMapper handling higher than 400, same thing it trickles down to the same error, I tried handling the 301, but if I return an exception obviously it fails, if I return null, then it goes back to the original exception.

So my questions would be, how does the redirect work, if 3XX (301 in my test) is considered a ClientWebApplicationException

[a comment by an automated account is not shown]

**Comment by michalszynkiewicz** [comment]
The property you mention is correct. Can you create a small reproducer for your problem?
Does changing the status code to e.g. 307 (temporary redirect) help?

**Reply by alex-martel** [reply]
Thank you for you reply. After a bit more digging, I see the redirect working fine for GET, in my case I was using a POST returning a 301, which then leads the the exception. I would assume this is by design?

**Reply by michalszynkiewicz** [reply]
So 301 works for GET but not for POST?

I found https://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.8 which seems to suggest this is the proper action...
TBH I wasn't aware that the behavior differs for POST.

**Reply by alex-martel** [reply]
Yes this is the behaviour I've been observing. At first I thought it simply did not work, but while making the reproducer with a GET it worked as it should, when changing the get to a POST then I get the exception. If you want I can push the reproducer if you'd like to play with it.

The link you posted indeed seems to suggest it is the proper action.

**Reply by pawel-ochrymowicz** [reply]
@manofthepeace, @michalszynkiewicz  I think configurable redirect handler could be a solution. 

At the moment quarkus rest client has no option to override `HttpClientImpl.DEFAULT_HANDLER` from underlying vert.x http client implementation. However vert.x exposes a method to override this logic: 
```java
@Fluent
HttpClient redirectHandler(Function<HttpClientResponse, Future<RequestOptions>> handler);
```
